Process Console command line history

chgenly opened the pull request:
This adds history to the ProcessConsole.  Use ctrl-up and ctrl-down to move through the history.

The code for the history has actually been added to the package for IOConsole.  However, I didn't want to change the behavior of IOConsole. The ProcessConsole constructor simply passes a flag to IOConsole to enable the history.  The existing constructor for IOConsole passes a false to a new version of the IOConsole constructor to disable history. Thus existing code using the existing constructor will still work as before.

The actual work for collecting history is in IOConsolePartitioner.

I fixed a bug in IOConsolePartitioner.computePartitioning.  There was a simple off by one type of error while looping through partitions to see which ones spanned the input region.  The loop was exited too early.  This fix is necessary for the history to work.

I fixed a bug in IOConsoleViewer.  When a user types input and the cursor is in a non-writable partition, there was code to insert the text into the beginning of the writable partition.  However it's more natural to add it to the end of that partition.  Consider the situation where the user has typed some input, perhaps "show info ". Then goes to a location earlier in the console to select a variable name and copy and paste it.  The user expects the pasted text to go after the text already typed.  That is they expect "show info var", not "varshow info ".  This fix is necessary for history.

In TextConsoleViewer, I changed the reveal job to optionally move the cursor to the end of the text widget.  This job is invoked if auto scroll is on.  When a document modification occurs at the end of the document, the cursor is also moved to the end of the document.  This is necessary for when a history line is placed in the document using ctrl-up or ctrl-down.

--
Capturing Keyboard Input from IOConsole

[Here is a very small change which could be separated into another branch.  I'm hoping this will be accepted in the same change as the history.

Perhaps someone knows of a better way to do this.
]

The constructor for IOConsoleInputStream has been made public.  This is the stream which receives keyboard input from a console once enter is pressed.  By making it public, a developer can set their own input stream.  This allows for logging, transforming and filtering user input before it is sent to the process. The only way IOConsoleParition will pass the text typed by the user to the stream is if it is an IOConsoleInputStream.

As background, IOConsole has a method to set the input stream, and
another to get the input stream.   getInputStream() returns
IOConsoleInputStream.  However, setInputStream() accepts the more basic
InputStream.  I'm not sure why the asymmetry.

vogella commented:
Project is currently in freeze mode due to the upcoming release. For the build to be successful you would also need to increase the y in the version number x.y.z.qualifier

vogella commented:
@mihnita could you review this contribution?

mihnita commented:
ACK. I'll review (in fact I've already started)

---

I also want to build and play with it a bit.

A few questions, to better understand what I'm looking at.

What does "history" means in the context of a console?
Because I am not sure I'm familiar with with the term, and I don't see any feature request, or a description of what that is.
And I was unable to find (google) what this is. So it does not sound like a common thing for consoles / terminals.
At least not with this terminology...

I am also a bit confused by the description mentioning inserting the text (anywhere other than the end).
I'm only getting familiar with Eclipse consoles, but I thought it is only possible to append to a console, not to insert?
Since the only way to interact with it is through input / stdout / stderr streams.
So there is no positioning.

Thanks a lot,
Mihai

laeubi commented:
The old bugzilla for this is here:
- https://bugs.eclipse.org/bugs/show_bug.cgi?id=561001

I have migrated this ti Github here:
- https://github.com/eclipse-platform/eclipse.platform/issues/532

chgenly commented:
> What does "history" means in the context of a console? Because I am not sure I'm familiar with with the term, and I don't see any feature request, or a description of what that is. And I was unable to find (google) what this is. So it does not sound like a common thing for consoles / terminals. At least not with this terminology...
> 

Sometimes commands typed into the console can be long.  If an error is made, it will need to be retyped.  A command line history feature makes this easy. Press ctrl-up arrow and the last command entered will appear. Now it can be edited, and finally enter can be pressed to reissue the command to the connected process. 

Sometimes a previously entered command needs to be reissued, just pres ctrl-up arrow  until you see the command.  If you pass it you can use ctrl-down arrow to move through the history in the other direction.

> I am also a bit confused by the description mentioning inserting the text (anywhere other than the end). I'm only getting familiar with Eclipse consoles, but I thought it is only possible to append to a console, not to insert? Since the only way to interact with it is through input / stdout / stderr streams. So there is no positioning.

There are two kinds of partitions.  Input and Output. Output partitions are read-only.  Input partitions are initially writable, once enter is pressed they become read only.  Writable partitions appear to only happen at the end of a document.  When an input  partition is writable, text can be inserted into it anywhere.  Typically when the user types, text is appended to this partition.  However, the user can use the left and right arrows to move anywhere on the input line.  Text can then be inserted at that position.

Pressing the ctrl-up and ctrl-down keys will replace the last writable input partition with a history line.

#

iloveeclipse reviewed (changes requested):
Please consider set API baseline & check API errors reported, there are plenty.

See 
![image](https://user-images.githubusercontent.com/964108/221158744-7f1ea1cf-4a71-4af2-a958-9712454974ad.png)

May be this could help: https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md

See also https://wiki.eclipse.org/Platform/How_to_Contribute#.5B4.5D_Prepare_API_tooling

chgenly commented:
I'll work on the baseline problems.  Thanks for pointing this out.

lbownik commented on org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsolePartitioner.java:
```diff
@@ -1465,4 +1478,65 @@ private void checkPartitions() {
+			return;
+		}
+
+		int offset;
```
i'd explicitely initialize this to 0 to convey te information that 0 is a legit value when this variable is used in line 1524

lbownik commented on org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsolePartitioner.java:
```diff
@@ -1465,4 +1478,65 @@ private void checkPartitions() {
+		}
+
+		int offset;
+		int length;
```
i'd explicitely initialize this to 0 to convey te information that 0 is a legit value when this variable is used in line 1524

lbownik commented on org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsolePartitioner.java:
```diff
@@ -1465,4 +1478,65 @@ private void checkPartitions() {
+				partitions.add(lastPartition);
+				inputPartitions.add(lastPartition);
+			}
+			offset = getLastPartition().getOffset();
```
what if getLastPartition() returns null?

chgenly replied:
This is already checked for at the line marked below.
```
			IOConsolePartition lastPartition = getLastPartition();
			// If lastPartition is null, there is no user input.
			// If lastPartition is read-only, the user has already pressed <enter>.
			// In either case, we need to create a new one to hold the input text.
			if (lastPartition == null || lastPartition.isReadOnly()) {     //<=============
				lastPartition = new IOConsolePartition(getDocument().getLength(), console.getInputStream());
				partitions.add(lastPartition);
				inputPartitions.add(lastPartition);
			}
			offset = getLastPartition().getOffset();
			length = getLastPartition().getLength();

```

lbownik commented on org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsoleViewer.java:
```diff
@@ -174,6 +177,20 @@ protected void handleVerifyEvent(VerifyEvent e) {
+		ITypedRegion[] regions = partitionerExt.computeWritablePartitions(startOfWritableParition, 0);
+		int endOfWritablePartition;
+		if (regions.length == 0) {
+			endOfWritablePartition = startOfWritableParition;
```
how about just
return startOfWritableParition;

chgenly replied:
Fine

chgenly commented on org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsoleViewer.java:
```diff
@@ -174,6 +177,20 @@ protected void handleVerifyEvent(VerifyEvent e) {
+			endOfWritablePartition = startOfWritableParition;
+		} else {
+			ITypedRegion region = regions[regions.length - 1];
+			endOfWritablePartition = region.getOffset() + region.getLength();
```
Fine

iloveeclipse commented:
Please look at test fails: https://ci.eclipse.org/platform/job/eclipse.platform.debug/job/PR-112/5/

chgenly commented:
Gah!  I accidentally deleted my commits while attempting to bring my master branch in sync with the upstream master.  I reconstructed my changes and have committed them to my local repository.  But the previous message "Pull request closed" indicates a problem.  Can a reopen this pull request, or do I have to create a completely new pull request.

jonahgraham commented:
You can reopen it.

Commit:
Accidentally deleted all commits on my branch.  This commit restores the
changes.  Plus it fixes unit test failures.

In the commits I lost, I was passing a parameter to the ProcessConole
constructor to enable command line history.  I determined this was not
the best way.  Now there is a method on IOConsolePage used to enable
command line history. Consider this scenario:
- User launches an application to be debugged.
- The launch delegate creates the process.
- The launch is added to the launch manager.
- ProcessConsoleManager listens for the new launch, creates
  a ProcessConsole, and adds it to the ConsoleManager.
  Notice the process console was created.  So there is no
  possibility of passing an enable command line history
  parameter to the constructor.
- The console view listens for ConsoleManager additions,
  which then creates a page for the console.  IConsolePageParticipants
  are called.
- The page participant is the place to enable command line history.

Created IIOConsolePage so command line history can be enabled.  The
history for a console has to enabled via the page so curosor management
can be performed.

The off by one error I mentioned earlier is not an error, the
original code was correct.

Add JUnit test for command line history.

chgenly commented:
Things should be stable again.  Unit tests pass on my local machine.

chgenly commented:
Things should be stable again.  Unit tests pass on my local machine.

laeubi commented:
This repository was merged with https://github.com/eclipse-platform/eclipse.platform if this PR is still relevant please rebase it and push it to the new origin!

chgenly commented:
I'm not sure what's involved.

Do I have to fork from the new repository first? Do I have to open a new
pull request against the new repository from the newly forked repository?

Can I open a pull request from my original repository to the new origin?


On Sat, Jul 1, 2023 at 4:21 AM Christoph Läubrich ***@***.***>
wrote:

> This repository was merged with
> https://github.com/eclipse-platform/eclipse.platform if this PR is still
> relevant please rebase it and push it to the new origin!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1615856662>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADWIH3SPUA7M5ESBP4MGFDXOAB4JANCNFSM6AAAAAAVCIBOCQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***
> com>
>

laeubi commented:
There are different techniques:

1. You can fork the new repository and create a new branch from master, then copy over your changes, review them and create a fresh PR
2. You can add the fork of the new repository to the old one and push the branch then to the new, after that rebase the branch ontop of the master but maybe need to resolve conflicts.

as there are only a few files maybe option 1 is the easiest one.

chgenly commented:
Thank you

On Mon, Jul 3, 2023, 9:08 PM Christoph Läubrich ***@***.***>
wrote:

> There are different techniques:
>
>    1. You can fork the new repository and create a new branch from
>    master, then copy over your changes, review them and create a fresh PR
>    2. You can add the fork of the new repository to the old one and push
>    the branch then to the new, after that rebase the branch ontop of the
>    master but maybe need to resolve conflicts.
>
> as there are only a few files maybe option 1 is the easiest one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1619444992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADWIH33POJ3TE2QQNXVLX3XOOJNJANCNFSM6AAAAAAVCIBOCQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***
> com>
>

chgenly commented:
I've used the oomp installer with the configuration provided by the launch
button on

https://www.eclipse.org/setups/installer/?url=https://raw.githubusercontent.com/eclipse-platform/eclipse.platform.releng.aggregator/master/oomph/PlatformSDKConfiguration.setup&show=true

This does the miraculous job of installing eclipse, cloning git
repositories and setting up the work environment.  It all builds!

Now I need to install the software I'm developing.  So I use the Runtime
Workspace launch.  Now I have a new eclipse; an eclipse launched by
eclipse.  Call it eclipse2.  I import my software into eclipse2.  It
doesn't build because I need various packages: guava, subclipse, cdt, mwe2
and so on.  Take mwe2 as an example, I install it using the 'help/install
new software' menu entry.  The install seems to go fine, but I don't have
any mwe2 support.  If I look in 'help/about/installation details' I can see
mwe2 listed as a feature.  However, when I go to plugins there are no mwe2
plugins shown.

Am I using the correct launch (Runtime Workspace)?  Am I using the proper
procedure to install software?

On Tue, Jul 4, 2023 at 8:55 AM Chris Genly ***@***.***> wrote:

> Thank you
>
> On Mon, Jul 3, 2023, 9:08 PM Christoph Läubrich ***@***.***>
> wrote:
>
>> There are different techniques:
>>
>>    1. You can fork the new repository and create a new branch from
>>    master, then copy over your changes, review them and create a fresh PR
>>    2. You can add the fork of the new repository to the old one and push
>>    the branch then to the new, after that rebase the branch ontop of the
>>    master but maybe need to resolve conflicts.
>>
>> as there are only a few files maybe option 1 is the easiest one.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1619444992>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AADWIH33POJ3TE2QQNXVLX3XOOJNJANCNFSM6AAAAAAVCIBOCQ>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***
>> com>
>>
>

laeubi commented:
Yes this is a (known?) problem that software installs does not persist restarts I think I also opened a bug somewhere but it seems not trivial to fix.

But what I'm wondering, why do you need all those extensions in "eclipse2"? I would assume one wants to develop such feature with a very minimal test-case and it should work without additional things...

chgenly commented:
Thanks for the reply.  I was doing it because I'm working on porting our
company's code to use the new feature.  I'm not sure how I'll do this now.
Perhaps I can build an eclipse with my modifications, and then test in that
environment.  How do I build Eclipse?

On Fri, Jul 7, 2023 at 10:34 PM Christoph Läubrich ***@***.***>
wrote:

> Yes this is a (known?) problem that software installs does not persist
> restarts I think I also opened a bug somewhere but it seems not trivial to
> fix.
>
> But what I'm wondering, why do you need all those extensions in
> "eclipse2"? I would assume one wants to develop such feature with a very
> minimal test-case and it should work without additional things...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1626869075>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADWIH4OCV7PZJ6EKADIUD3XPDWOJANCNFSM6AAAAAAVCIBOCQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***
> com>
>

merks commented:
FYI. When I want additional things in the debug launched IDE I include them in the target platform and include those in the launch configuration.

chgenly commented:
Thanks, I'll give that a try.

On Mon, Jul 10, 2023 at 10:35 AM Ed Merks ***@***.***> wrote:

> FYI. When I want additional things in the debug launched IDE I include
> them in the target platform and include those in the launch configuration.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1629414888>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADWIH5WSX2YC7FR3NKWLVTXPQ4PBANCNFSM6AAAAAAVCIBOCQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***
> com>
>

chgenly commented:
I created a pull request for command line history.  What do I have to do
next?  How do I associate the pull request and the bug report? I believe
this is the correct bug report:
https://github.com/eclipse-platform/eclipse.platform/issues/532

On Mon, Jul 10, 2023 at 10:51 AM Chris Genly ***@***.***> wrote:

> Thanks, I'll give that a try.
>
> On Mon, Jul 10, 2023 at 10:35 AM Ed Merks ***@***.***>
> wrote:
>
>> FYI. When I want additional things in the debug launched IDE I include
>> them in the target platform and include those in the launch configuration.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/eclipse-platform/eclipse.platform.debug/pull/112#issuecomment-1629414888>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AADWIH5WSX2YC7FR3NKWLVTXPQ4PBANCNFSM6AAAAAAVCIBOCQ>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***
>> com>
>>
>
